### DISCUSSION TITLE
Expression of type Dict[Any, Any] cannot be assigned to TypeVar bound to Dict[Any, Any] if also bound to Any?

### DISCUSSION THREAD

**Discussion by ringohoffman** [microsoft/pyright#5199] [Q&A] [answered]
For reference, I'm trying to add some type hints [here](https://github.com/pytorch/pytorch/blob/e380d692dc2c20d711c2155fc580eec30383d301/torch/optim/optimizer.py#L397), and I am seeing some surprising behavior that only seems to happen if the `TypeVar` is also bound to `Any`:

```python
from typing import Any, Dict, TypeVar

S = TypeVar("S", int, Dict[Any, Any], Any)

def cast(value: S) -> S:
    if isinstance(value, int):
        return value
    if isinstance(value, dict):
        return {k: cast(v) for k, v in value.items()}  # Expression of type "dict[Any, Any]" cannot be assigned to return type "S@cast"
    return value


R = TypeVar("R", int, Dict[Any, Any], object)

def cast2(value: R) -> R:
    if isinstance(value, int):
        return value
    if isinstance(value, dict):
        return {k: cast2(v) for k, v in value.items()}  # OK
    return value


T = TypeVar("T", int, Dict[Any, Any])

def cast3(value: T) -> T:
    if isinstance(value, int):
        ...
        return value
    return {k: cast3(v) for k, v in value.items()}  # OK

```

I am a little confused why this is happening. Is this intentional? If so could you explain?

**Comment by erictraut** [comment]
These are constrained TypeVars, not bound TypeVars. Constrained TypeVars are unique to Python and were added to the Python type system primarily to handle `AnyStr`. I recommend against using them in most cases.

A constrained TypeVar is meant to constrain the type to one of the specified types. Ideally, all of these types are distinct and not overlapping (i.e. one is not a subtype of another). PEP 484 doesn't specify what happens when one of the constraint types is a subtype of another. I've added some logic to deal with this case, but you're in "undefined" territory if you use this.

Adding `Any` to a constrained TypeVar type list is a very unusual thing to do. It's not clear what that would even mean because `Any` is both a subtype and a supertype of all types. 

Perhaps if you explain what you're trying to do here, I can offer a better solution.

**Reply by ringohoffman** [reply]
My goal is to type `cast` as a function that returns the same type as its `value` parameter, but I can't seem to get pyright to accept that a new`dict` constructed from the `items` of `value` is the same type as `value`.

```python
from typing import TypeVar

S = TypeVar("S")

def cast(value: S) -> S:
    if isinstance(value, int):
        return value  # OK
    if isinstance(value, dict):
        reveal_type(value)  # Type of "value" is "dict[Unknown, Unknown]*
        return {k: cast(v) for k, v in value.items()}  # Expression of type "dict[Unknown, Unknown]" cannot be assigned to return type "S@cast"
    return value
```

My goal with using `Any` is to handle the default case when none of the conditions are met so that value is not typed as `Never`:

```python
from typing import TypeVar

S = TypeVar("S", int, str)

def cast(value: S) -> S:
    if isinstance(value, int):
        return value
    if isinstance(value, str):
        return value
    reveal_type(value)  # Type of "value" is "Never"
    return value
```

Isn't everything an instance of `object` as well? So why does that not trigger the same error that `"dict[Unknown, Unknown]" cannot be assigned to return type "S@cast"`?

**Reply by erictraut** [reply, marked as answer]
> Isn't everything an instance of object as well?

Yes, all types are subtypes of `object`, but the converse isn't true. The `Any` type is special in that it's effectively both a "top" and a "bottom" type, so it has a bunch of special-cased behaviors. Using `Any` as a constraint in a constrained TypeVar doesn't really make sense, so you shouldn't be surprised if you run into odd behaviors here. I don't consider this a bug in pyright; it's an invalid use of a typing feature.

An overload would be a much better solution to the problem you're trying to solve here. You have a polymorphic function whose return type depends on the input, and this is what `@overload` is intended to handle. 

```python
from typing import Any, TypeVar, overload

S = TypeVar("S")

@overload
def cast(value: int) -> int: ...
@overload
def cast(value: dict[Any, Any]) -> dict[Any, Any]: ...
@overload
def cast(value: S) -> S: ...

def cast(value: int | dict[Any, Any] | S) -> int | dict[Any, Any] | S:
    if isinstance(value, int):
        return value
    if isinstance(value, dict):
        return {k: cast(v) for k, v in value.items()}
    return value
```

**Reply by ringohoffman** [reply]
I have to say I'm not really satisfied by a `dict` constructed plainly from another `dict` not being able to be bound to the same type when coming from a `TypeVar`:

```python
from typing import TypeVar

T = TypeVar("T")

def f(obj: T) -> T:
    if isinstance(obj, dict):
        return {k: v for k, v in obj.items()}  # Expression of type "dict[Unknown, Unknown]" cannot be assigned to return type "T@f"
    return obj

def g(obj: T) -> T:
    if isinstance(obj, dict):
        return dict(obj)  # Expression of type "dict[Unknown, Unknown]" cannot be assigned to return type "T@g"
    return obj


x = dict({1:2, 3:4})
reveal_type(x)  # Type of "x" is "dict[int, int]"

if isinstance(x, dict):
    x = {k: v for k, v in x.items()}
    reveal_type(x)  # Type of "x" is "dict[int, int]"
```

Why can `dict[Unknown, Unknown]` not be treated more like `dict[Unknown1, Unknown2]` s.t. you can still track the types of the keys and values? It seems unintuitive that it works generally except for unbound/unconstrained `TypeVar`s.

IMO in the above example there shouldn't be any reason the signature wouldn't be able to just be

```python
def cast(value: S) -> S: ...
```

**Reply by erictraut** [reply]
The error you're seeing here isn't related to the fact that the type arguments for `dict` are `Unknown`. The problem is that `dict` (regardless of which type arguments it has) is not compatible with type variable `T`. You have used an `isinstance` check to determine whether `obj` is a subclass of `dict`, but that doesn't mean that it's actually a `dict`. It could be an instance of some class that derives from `dict`, so a type checker cannot assume that it is compatible.

Conditionally checking for specific types of a generic parameter is an anti-pattern. If your function is generic, it shouldn't care what type was passed. Your use of `isinstance` to determine the type of a generic parameter is a red flag that you're doing something a bit wrong here.
